### DISCUSSION TITLE
Draft design doc: Beatmap of the day

### DISCUSSION THREAD

**Discussion by bdach** [ppy/osu#27983] [Ideas] [answered]
This is a design document of sorts for the feature floated today on discord (see [discussion here](https://discord.com/channels/188630481301012481/188630652340404224/1232587729246748714) and then continued [in thread](https://discord.com/channels/188630481301012481/1232593041416786002)).

Most things below have already been brought up in the discussions above but some of it is explicitly my proposal. Assume everything in here is negotiable / up for debate.

# Basic premise

- Every day a single beatmap difficulty is selected as the "beatmap of the day".
- The beatmap of the day gets a playlist created for it, which is open for 24 hours.
- Beatmaps to pick can be either queued manually, or chosen automatically.
  - Automatic choice criteria: beatmap is from a featured artist, beatmap is in a specified star rating range, no repeats if possible?
- The beatmap of the day is accessible from the game's main menu, next to solo / playlists / multiplayer.
- The beatmap of the day screen displays real-time statistics and leaderboards as user submit scores, and includes a chat box.

## Ideas for future expansion

- Individual rewards for top N players on every playlist.
- Collective rewards for everyone for reaching selected targets.
   Ideas for targets:
    - Reach collective `$playcount` in the playlist.
    - Collect `$total_score` cumulatively.
    - Get `$total` A/S ranks or above.
    - Have `$count` of players achieve accuracy of `$percentage` or above.
    - Have `$count` of players reach `$maxCombo`.

## Tentative UI mockups

https://www.figma.com/file/4oaiT5sBCiD2cLjfaEfv1U/Untitled?type=design&node-id=0%3A1&mode=design&t=kd5c0i2eW7vf5Aw3-1

Very makeshift, cobbled together using various scraps from @arflyte's various figmas over the years.

Please ignore the fonts being all over the place. I can't be bothered with figma bullshit. And please don't focus on this too much to begin with, the details of appearance are secondary to what this system is supposed to be doing.

# How to execute this

- Something must create the daily playlist on a cron.
  - Choices: `osu-web` or `osu-server-spectator`.
- A beatmap queueing system must exist for manual queueing. Probably something based on a database table. It should be hooked up to ASS so that it can be managed without our direct intervention as devs.
- `osu-server-spectator` should announce the daily playlist to all clients.
- Score submission will use the existing playlist flow / infra.
- Real time stats updates for the daily playlist will be delivered to clients by `osu-server-spectator`. It'll probably be `osu-server-spectator` querying most relevant stats in a loop on an interval and sending that data to all clients at once (there should be no reason to send different data to different clients in the simplest variant).

## How to actually decompose this

1. Implement the queueing mechanism and the room creation cron.
2. Add the button in client, initially using the playlists screen unmodified.
3. Replace the playlists screen with the new one that displays the real time info.

# Open questions

- How to include the biggest number of players possible
  - Skill levels: probably will be addressed by picking different star rating ranges on each weekday
  - Rulesets: this is an open one, maybe we'll run mostly osu! to begin with to see if the system takes, then try other rulesets for a single day every now and then, and if interest keeps up, maybe we can expand the system to have 1 beatmap per ruleset or something
- Rewards
  - Many ideas thrown out, including
    - medals/badges (probably not viable)
    - granting supporter
    - announcement in chat
    - temporary player title
    - ability to suggest next beatmap
    - some sort of new display
    - tying into a revamp of XP somehow

**Comment by Sleepteiner** [comment]
I do like how this proposal will encourage players to engage in the competitive side of osu more regularly and in more varied ways, however I worry that having the playlists only be open for 24 hours will make it impossible for a lot of players to participate as consistently as they would like to. Because of this, instead of a "beatmap of the day" system, I think it would be better to have a "beatmaps of the week" system.

Instead of picking a map every day and encouraging players to compete on that map for only 24 hours, we could pick 5 (or so) maps every week and keep the playlist of those maps open for the whole week. This way players who are only able to play the game on the weekends will still be able to participate and compete just about as much as players who have the time to play daily.

I understand that this will create something of similar appeal to the current Beatmap Spotlights system, so maybe the Spotlights project should just be merged into the hypothetical "beatmaps of the week" system so that the players who enjoy the Spotlights have something to look forward to each week consistently and the people running the Spotlights project no longer have to deal the the constant issues of trying to make a 9 week competition flow smoothly in a client that is rapidly changing.

**Reply by peppy** [reply]
> Instead of picking a map every day and encouraging players to compete on that map for only 24 hours, we could pick 5 (or so) maps every week and keep the playlist of those maps open for the whole week.

I don't think this creates the same kind of dynamic I'm hoping for – maps-per-week just feels very.. static? Having one a day means there's something new at the main menu each day, and something to look forward to. Maybe just me though.

Users don't need to play every day, and they can always go back and look at previous days they missed after the fact.

**Comment by felipemarins** [comment]
Overall, I like the idea and how it was proposed here. My only concern for now is about the individual rewards. As it is described, the idea is to give some kind of reward to the top N players. As I see it, it could be a problem, especially if you want to have less skilled players competing for it, as it could turn the lower star rating days into a competition between more skilled players for who gets the maximum score first as soon as the playlist starts. So maybe the individual rewards should be limited to temporary things only visible that day, as it would be less likely for something like this to happen this way.

**Reply by peppy** [reply]
I was hoping for rewards which aren't necessarily "top x" – maybe something for getting a certain rank (maybe for chaining over multiple days?)

**Reply by felipemarins** [reply]
> maybe something for getting a certain rank (maybe for chaining over multiple days?)

Yeah, I would have no problem with that. Just to be clear, I don't think it's something bad to give top players a reward, I was just worried about the impact it could have in the lower star rating days if the individual rewards were limited to that, as the purpose of those days would be to let less skilled players to have some fun with this new system.

**Comment by Bubbleman532** [comment]
This needs more thought put behind how it will run before an implementation is attempted - if a map has multiple difficulties, should a player get an S rank on each before being able to set a score on the next one up? Are mods allowed, restricted, enforced or fully disallowed, and does this change per mapset per day? Is a leaderboard open for 24 hours, or are people given half an hour to set the best score at set times of the day (similar-ish to cup of the day)?

That said, I and others would prefer to see the song select redesign prioritised before anything new is added to the game, as well as a results screen redesign and a rework to the first-time setup. These three things currently are the biggest barriers to use presented to stable players.

I have a further roadmap I've written up past these initial points if it's possible to present it to devs and discuss the reasonings.

**Reply by peppy** [reply]
> if a map has multiple difficulties, should a player get an S rank on each before being able to set a score on the next one up?

In discussion, we were proposing one difficulty only per day. Potentially ramping up over the week from easy (say 3 stars) to hard (say 6 stars).

> Are mods allowed, restricted, enforced or fully disallowed, and does this change per mapset per day?

These would be mixed up to create more diversity.

>  Is a leaderboard open for 24 hours, or are people given half an hour to set the best score at set times of the day (similar-ish to cup of the day)?

Likely 24 hours or something close to it (maybe 18 to give a break before the next one).

> This needs more thought put behind how it will run before an implementation is attempted

The questions you've raised are things that can be tweaked as we go based on reception. The important part here is getting the basics in place so we can run the event in the first place 😄 

> That said, I and others would prefer to see the song select redesign prioritised before anything new is added to the game, as well as a results screen redesign and a rework to the first-time setup. These three things currently are the biggest barriers to use presented to stable players.

Well aware of these points, and work is continuing on those (in theory). This is a proposal for the next new thing in lazer, which will run in parallel to other efforts.

**Reply by lucasmelodev1** [reply]
> These would be mixed up to create more diversity.

Idea: prioritize "Fun" type mods over normal ones to give people with "strange" talents to shine. This idea is based on my fear of having the same players on the top always

**Comment by smoogipoo** [comment]
I think that the average player is not going to play one beatmap for a day, and I'll wager that even a player capable of getting top-1 on any beatmap will not play one for a day. osu! players fall in two categories: either they're motivated to achieve something in which case they'll retry-spam, or they're pretty short-attention-spanned/playing for fun in which case they'll prefer variety.

Let's say a beatmap is picked that the average player can play and doesn't mind listening to on repeat, which is the best case scenario. A top player comes along and easily gets 1,000,000 points (max score). What:
- Motivates the top player to continue playing the beatmap?
- Motivates any other player to continue playing, if they can never surpass them?

I think that it's crucial to answer for these two questions with the initial implementation, as not doing so will take the fun out of the system.  
Trackmania addresses both of these by having phases and divisions, and I'd propose the same system here:
- Have a qualification phase that starts at an exact UTC time of day. This will motivate players to consistently log onto lazer a little bit just to qualify. This could also be announced on the website in the form of a countdown, with the goal being to give players something to _actively_ look out for and schedule in their time. Countdowns will likely appear in players' discord servers, etc.
- Separate players into divisions where you don't have to feel like you suck and won't get any rewards when you get 400k points versus top players at 900k+.
- Have a knockout phase that ends fairly quickly - players by this point have already been playing the map for a considerable amount of time in the qualification phase. At this point even top players that previously achieved the max score will have to prove their consistency in order to not get knocked out.
- Naturally, this means that the mini tournament doesn't run for 24 hours, and I don't think we want that anyway. Rulesets should be staggered to allow competing in multiple ones - I think it could motivate players to try out other rulesets if you can earn awards doing so regardless of your skill level/division.

I see this being done more by the multiplayer server than playlists.

**Reply by bdach** [reply]
To preface the rest of this comment, my goal is not to shoot this down or anything, I'm not that attached - but what you describe would be a completely different system, with completely different considerations - and a different goal. While what the OP is about is more of a "try something new today" / possibly community cooperation vibe, I feel like your thing is more geared towards individual competition. Whether either of these is more desirable I'm not sure I would be able to tell.

My main questions if aiming to develop this into a full-fledged proposal would be:

- How long are the gameplay loop cycles here? You mention it not being "24 hours", but how long would the qualifiers/knockout have to last to not exclude players due to getting timezoned? Also note that an actual knockout phase is an inherently serialising/real-time commitment. I imagine that all players that qualify must be online to play it. If it gets too long it'll exclude everyone but the most committed.
- How does the format work? How many divisions, what's the knockout phase like? Is it a bracket? If yes, is it single-elim, double-elim, whatever?
- How do we run a knockout phase using the current multiplayer system to begin with? It sounds that if we wanted to use the existing system it would require either one mp room per match, or I guess serialising the matches in a queue in a single room?

**Reply by peppy** [reply]
I'd definitely consider this as a follow-up effort because there's a lot more moving pieces (mentioned in the original thread that any realtime mode would be a separate effort). Let's start with the simple version because we can get it live quickly and iterate based on how much people like it.

> Motivates the top player to continue playing the beatmap?

I don't think there necessarily needs to be motivation. It would be fine if the community goals only counted one play per person – the idea would be to get more people in there, imagine people sharing with friends or in public chat to try and get to the goal.

> Motivates any other player to continue playing, if they can never surpass them?

In the thread on discord we discussed having some kind of little star on the profile for just participating, or getting a certain (lower) grade like an A. And potentially a "streak" mechanism for the longest you've gotten the star in a row.

I want to keep this really *really* simple and get it out. This worked really well in yesteryear and means we can more more quickly. We can add more things to the system like tiers, qualifiers, multiplayer realtime addons depending on reception (or lack of).

> I think that the average player is not going to play one beatmap for a day

I know I would for one, and I hardly play osu!. It's something which would motivate me, so I see other casuals also being interested. The only catch is it needs to be in my skill range (but I'd be happy to skip days that aren't) 😉

**Reply by smoogipoo** [reply]
Yeah I sort of gathered that it was totally different from your intentions about 3/4 of the way through, but the initial proposal did start at CoTD. My worry is still that without a solid motivation structure, it'll just become a system that exists out-of-sight out-of-mind.

Perhaps I leaned too far into CoTD, for which I'll delegate to the rest of the discussion above.

---

To answer your questions:

> How long are the gameplay loop cycles here?

This is going to be a ballpark just to illustrate, but could be adjusted:
- Assuming a 3 minute beatmap.
- 1 hour qualification stage. This would be tweaked, with the goals of:
    1. Early birds should be able to qualify and do something else (have dinner/etc).
    2. Those playing for the entire duration should have a chance to get the best score possible.
    3. Late birds should have a chance to qualify themselves, basically the inverse of early birds.
- 10 minute break.
- 20 minute knockout stage. Assuming the top-8 division, where one players gets knocked out every run. Other divisions would scale proportionally: top-16 -> 2 players knocked out every run, top-32 -> 4 players, etc...

The entire thing would last for around 1 - 1.5 hours, and I think it's important for the knockout stage to directly follow.

Some players will inevitably get "timezoned", or have IRL commitments, or perhaps even have other commitments within osu!. I think that's fine but we'd also have to play that by ear - I've never scheduled tournaments so I can't talk from experience.

Perhaps we can do even-odd weeks but I think it's important to have some sort of schedule. The goal shouldn't be to have everyone playing every day, but a consistent schedule that a player and their friends can look forward to.

> How does the format work? How many divisions, what's the knockout phase like? Is it a bracket? If yes, is it single-elim, double-elim, whatever?

It's more like group stages? See my above example. You could have as many divisions as you have reward levels. I kinda like TM's model of different trophy levels, that you could display on the profile as a list of trophy level + number of times achieved combo.

I wanna point out also that TM's divisions aren't top-8, top-16, etc, but top X%. I think that's perfectly fine. It pretty much follows what a division is in traditional matchmaking systems - those that have bronze/silver/gold/platinum/diamond/master/grandmaster for example.  
Those could even be the rewards - you get a time-limited badge (expiring every month, let's say) on your profile that says "i'm bronze 5". Maybe that becomes a way to seed players in normal tournaments that currently use a [badge-weighted system](https://osu.ppy.sh/wiki/en/Tournaments/Badge-weighted_seeding), but I don't want to dream too far ahead.

> How do we run a knockout phase using the current multiplayer system to begin with? It sounds that if we wanted to use the existing system it would require either one mp room per match, or I guess serialising the matches in a queue in a single room?

I'd really want to use a single room.

**Reply by Bubbleman532** [reply]
Smoogi has the right idea here, the format should be a timed competitive experience that everyone can take part in as a community, but focusing on a map of the day and then building towards cup of the day should probably be the focus.

>>How do we run a knockout phase using the current multiplayer system to begin with? It sounds that if we wanted to use the existing >>system it would require either one mp room per match, or I guess serialising the matches in a queue in a single room?
>
>I'd really want to use a single room.

Qualifiers could be hosted in one room with one leaderboard, where the top 64 and each subsequent group then get moved into their own room once knockouts start - To prevent repetition, I played around and landed with 64 - 48 - 36 - 24 - 16 - 10 - 6 - 4 - 2 - 1 as the progression, with spectator being forced on for those players who are knocked out to prevent them from playing again. Given a 3 minute map, this is 9 runs, meaning 27 minutes of play (with maybe a 30-60 second break between rounds?). I think a 30 minute qualification would be a good middle ground between having time to set a score and improve it, and having time to rest before the knockouts.

Divisions could also be handled with lower star rating beatmaps, perhaps requiring a player to play up the difficulties on a mapset on their first time playing MotD, needing an S rank to move up to the next difficulty on a map and being restricted from moving back down - this comes with the caveat that it could promote sandbagging but I'd hope that issue doesn't come up or can be solved by someone who expands this idea. Players could then be allowed to skip to the highest difficulty that is below their peak star rating FC on a previous day's map.

I also don't think picking only unique maps is the way to go - picking a 'generic' but well made map will appeal to casual players who don't lean towards unique technical experiences.

>I wanna point out also that TM's divisions aren't top-8, top-16, etc, but top X%

TM does use a top 64 division 1, with each subsequent 64 being 2, 3 etc. Having a player play the hardest map on a set that they can could work in a similar way, where top 64 on the hardest difficulty are division 1, which when a division isn't completely filled progresses to the top 64 on the 2nd hardest difficulty starting off the next division down.

**Reply by Sleepteiner** [reply]
The more time limited and the more of a time commitment this becomes, fewer and fewer players will have the ability to compete. If fewer players compete, the project will become less competitive. The less competitive the project is, the motivation for players to regularly put in the time necessary to participate will go way down.

Maybe I'm wrong about all of this since there isn't any large scale precedent for this kind of project in the osu community that I know of, but I really can't see a "cup of the day" system being much more than a novelty that the average player might engage with a few times before simply ignoring since the level of competition likely won't be too high on average and taking part would require both scheduling and a time commitment on their part. At least with the basic "map of the day" system, players will be regularly encouraged to at least casually participate in the competitive side of the game because lots of other players will likely take part and all someone has to do to compete is open the game and play the map once.

> I think that the average player is not going to play one beatmap for a day

True, but that's not what the vast majority players would be doing anyway. The vast majority of players would just play the map until they get a score they are satisfied with and then move on.

> Motivates the top player to continue playing the beatmap?

They don't need motivation to keep playing after getting the max score, they can just move on to the rest of the game. Plus, if we allow all score increasing mods, it would be very rare for someone to get the max score unless maps under 3 stars are picked frequently.

> Motivates any other player to continue playing, if they can never surpass them?

If two players get the max score on a map, they could just tie for first place, meaning the motivation to get first place would still be in effect.

**Reply by lucasmelodev1** [reply]
I like the idea of the divisions in the first iteration of the beatmap of the day. Me, as a 5 digit player, would love to compete only with 5 digit players (or even low-5-digits). If I were to compete with everyone, I think I would not be motivated to play in any day. It's just me, though.

**Comment by peppy** [comment]
Some additional notes/thoughts:

> Beatmaps to pick can be either queued manually, or chosen automatically

I'd hope that we don't need to resort to automatic selection of future beatmaps. I do think in place of any automation, there should be some kind of alerting when the queue is empty for the next week period so someone can jump in and fill it up. I've raised this requirement internally and there is initial interest in taking on this role, so that's good.

> Something must create the daily playlist on a cron

I think osu-web is more suited for this task. It has the ability to message us on discord ops channels if things go wrong, and has a cron system in place already which (in most cases) works reliably.

> Rewards

It'll be good to have some kind of records of participation on user profiles (probably using a new display style since existing medals/badges don't allow for more-than-one achievement of same thing). I'm more interested in participation and streaks than winners, because i think people will get more enjoyment from that (but we won't forget winners either).

---

Regarding play statistics – while we will initially use them for this special mode, I see no harm in recording stats for all playlist rooms. It's something we can potentially display on user playlists in the future and have them benefit from more "flavour" displayed on the playlist interface.

**Reply by Sleepteiner** [reply]
(I apologize for adding my comment so close to the release of this project.)

Personally, I am against the streak idea for the daily challenge project. Since a lot of people won't be able to play every day, having the game encourage them to do so anyway by displaying a day streak will just add stress which isn't fun. An example of this effect can be seen by looking at Duolingo communities. People in those communities have always frequently complained about how, for a big portion of the dedicated userbase, the streak system in Duolingo adds more stress than anything else.

I think it would be much better to have the statistic of "total days participated" instead of "number of days participated in a row" so that when someone who is normally very active can't play for a few days, the displayed statistic will simply not go up instead of resetting to 0.

**Reply by peppy** [reply]
Thanks for the feedback, we're going to be showing total days along with weekly streaks, which should satisfy people that can't play every day.

**Comment by bdach** [comment, marked as answer]
Further work on this will be tracked in https://github.com/ppy/osu/issues/28136.

**Comment by nikunaku** [comment]
Hi, I'm not sure where I'd put a suggestion like this, so I'll just put it here.

"Daily challenge" could maybe be renamed to _"dailies"_ or _"dailys"_.
This naming fits better in the menu it's in (singleplayer, multiplayer, playlists), being a single word. Additionally, terms like _playing dailys_, _daily leaderboard_, etc, are aesthetically quite pleasing.

**Reply by bdach** [reply]
I'm not renaming the feature again at this stage unless there's significant push for it. "Dailies" may well become common vernacular in player circles but I find it a bit inscrutable as a primary name for the feature because it could mean anything.

**Comment by iwa** [comment]
doesn't seem to have been discussed yet, but what time will be used for the daily reset?
most games' dailies have a different reset time for each region (US, Europe, Asia), will this daily challenge system use a unique time for everyone or adjust to at least those 3 regions so all players around the globe can get a sense of a "true" daily challenge?

**Reply by bdach** [reply]
It's midnight UTC right now, globally. I didn't really plan on "regional" adjustments.

With a runtime of 24 hours I'm not sure why this would matter.

**Reply by iwa** [reply]
well it just gives the players a true sense of a daily challenge, as it begins when day begins and ends when days ends
definitely not a requirement tho, i totally get that would be a lot of trouble to work with a regional system, but i feel some people may bring it up

edit: after thinking about it, people could get mad cause that would mean if there was any kind of rewards, US & Asia players could stack up 2 daily rewards in 1 session, but i feel that may not apply as much to osu as rewards wouldnt help people getting better as the game

**Reply by bdach** [reply]
> if there was any kind of rewards, US & Asia players could stack up 2 daily rewards in 1 session

Only the first time no? At most it would mean that these players would only need to play every other day rather than every day to keep up a streak. But that's something that can happen even with regional adjustments.

**Reply by iwa** [reply]
yes but that would feel like a unfair advantage if it gave rewards that would impact the progress/the gameplay (like a gacha game), as US & Asia would in fact only need to play every other day and still benefits of the rewards
but as that doesn't apply on osu (rewards wont impact player performance), it's actually not really a concern
still might may be brought up but yeah definitely not a requirement

**Comment by iwa** [comment]
don't know if you'd prefer me to open a dedicated discussion, lmk

what are the conditions for a day to be counted in the streak? it seems to be that you have to clear at least one time the daily map, right?

if that's the case, would it be able to always have NF in freemods?

don't know if it was intentional but yesterday (30/07) NF wasn't part of the choosable freemods
i feel that's a shame cause as a casual player, i struggled with the map (tracable mod was an horrible experience to me ahah......) and i was about to lose my streak just because of that

i get that you have to clear at least once the current map, totally makes sense, but it creates some difficulty imbalances with some mods

when comparing to other daily system, they are definitely easier in a sense where they adapt to the user:
- hoyo games, dailies are quite easy, in genshin for eg enemies levels are adapted to your level, so it always feels balanced
- duolingo, just need to complete a lesson a day so it follows your learning path
- trackmania, as long as you complete the map it's fine (afaik)

obviously, it's not possible to make uniques and adapted dailies for everyone, so always having NF as an option to let people who care about streaks being able to keep theirs seems like a good middle ground.

i feel that would def be more impacting when you'll implement other modes, if all modes uses the same queue, a lot of players won't be versatile enough to easily clear a 4* in every modes

**Reply by peppy** [reply]
we will probably have something like relax or halftime in freemod for those that need an option.

**Reply by iwa** [reply]
would def helps, thanks!
